### PR TITLE
Invalid order events treatment

### PR DESCRIPTION
**Pull request by squadgazzz**
# Description
Closes pt2 of the issue https://github.com/cowprotocol/services/issues/1977:
> special treatment for invalid events (maybe only store another invalid event when the order was previously NOT invalid; meaning only store state changes)

# Changes
- [ ] Insert `Invalid` order event if the latest order event label **is not** `Invalid`.

## How to test
Covered with an e2e test.

### CODE REVIEW

File: crates/e2e/tests/e2e/tracking_insufficient_funds.rs
Comment (squadgazzz) [inline crates/e2e/tests/e2e/tracking_insufficient_funds.rs:156]: This test doesn't pass on the `main` branch.

File: crates/e2e/tests/e2e/tracking_insufficient_funds.rs
Comment (squadgazzz) [inline crates/e2e/tests/e2e/tracking_insufficient_funds.rs:139]: This test doesn't pass on the `main` branch.

File: crates/database/src/order_events.rs
Comment (squadgazzz) [inline crates/database/src/order_events.rs:90]: In staging, this query takes around 60ms.

File: crates/e2e/tests/e2e/tracking_insufficient_funds.rs
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/tracking_insufficient_funds.rs:25]:
```suggestion
    let [trader_a, trader_b] = onchain.make_accounts(to_wei(10)).await;
```

File: crates/e2e/tests/e2e/tracking_insufficient_funds.rs
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/tracking_insufficient_funds.rs:168]: nit: Slightly verbose 😅 
```suggestion
    !events
        .windows(2)
        .map(|w| (w[0].label, w[1].label))
        .any(|window| window == (OrderEventLabel::Invalid, OrderEventLabel::Invalid))
```

File: crates/autopilot/src/solvable_orders.rs
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs:203]: I like your change to collect `OrderUid`s into different buckets depending on the label.
This can be applied to `filtered_orders` as well to make the code less verbose.

File: crates/database/src/order_events.rs
Comment (MartinquaXD) [reply crates/database/src/order_events.rs:90]: This is concerning because that doesn't bode well for prod. 😬 
Currently we discard around 520 orders from every auction because the owner does not have enough balance. If we go down this path we should probably spawn a background task for this insert as well.

**Review by MartinquaXD — APPROVED** [review]
Code looks good. 👍 
I'm still a bit worried about the 60ms per order event insert.
Maybe we should consider purging the databases from all duplicated `Invalid` events before merging?

File: crates/autopilot/src/solvable_orders.rs
Comment (fleupold) [inline crates/autopilot/src/solvable_orders.rs:235]: Oh wow, the code is quite lavish with the reasons for which it excludes orders. Would love to see all of those go away (and be done in the driver eventually). This has nothing to with this PR of course.

File: crates/autopilot/src/solvable_orders.rs
Comment (fleupold) [inline crates/autopilot/src/solvable_orders.rs:255]: Shouldn't we also store just state changes for `Filtered` orders? What's the point of seeing multiple Filtered events in a row? This would make the insert code more consistent.

File: crates/database/src/order_events.rs
Comment (fleupold) [reply crates/database/src/order_events.rs:108]: Can you execute the query like this in prod to test? Not sure if this helps, but we could add another boolean column in the table "latest" and then for every insert first update the latest record to be false. An index on this column could make lookups much faster.

**Review by fleupold — APPROVED** [review]
LGTM

File: crates/database/src/order_events.rs
Comment (MartinquaXD) [reply crates/database/src/order_events.rs:108]: The query used in this PR is constant time since we already have a index on `(order_uid, timestamp)`.
It simply scans the index for the given `uid` backwards and because we are only interested in the last row we only do a single iteration during the scan.

File: crates/autopilot/src/solvable_orders.rs
Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:255]: Could not analyze prod data since SQL queries are a bit complex to find occurrences of subsequent `filtered` labels within a single order. Staging data was used to calculate the ratio of subsequent events with the same label, while prod data was used to determine the ratio between events with different labels:
1. There are around 45% of `invalid` events where ~99% are subsequent, meaning the table size is expected to be decreased by ~45%(14GB).
2. Around 22% of `filtered` events will remain in the table and ~70% are subsequent. So, approximately ~15% could be additionally removed(2.5GB) from the remaining data.

In summary: 45% or reduction with `invalid` retreatment only and 55% including `filtered`.

File: crates/autopilot/src/solvable_orders.rs
Comment (fleupold) [inline crates/autopilot/src/solvable_orders.rs:236]: Do we still need the old method (I feel like all order events should only be updated if non-subsequent). If so, the method could also be called more neutral (ie, keep `store_order_events` but make it only store non subsequent events)

File: crates/database/src/order_events.rs
Comment (fleupold) [inline crates/database/src/order_events.rs:114]: nit: why not use the same variable?

**Review by fleupold — APPROVED** [review]
Looks good. I'd still slightly prefer if all order events are handled the same way, but not blocking this PR.

File: crates/autopilot/src/solvable_orders.rs
Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:236]: It was already mentioned here: https://github.com/cowprotocol/services/issues/1977#issuecomment-1833830663. And I think it was additionally discussed with @MartinquaXD. Filtering other labels such as `Ready`, `Traded`, `Considered` extends the time period between the events. If that won't harm metrics/debugging, we can apply this logic to all the events.

File: crates/database/src/order_events.rs
Comment (squadgazzz) [reply crates/database/src/order_events.rs]: Removed

File: crates/autopilot/src/solvable_orders.rs
Comment (fleupold) [reply crates/autopilot/src/solvable_orders.rs:236]: > Filtering other labels such as Ready, Considered extends the time period between the events.

Do you see any way in which this could make a difference? If I understand correctly `ready, considered, considered, traded` would become `ready, considered, traded` which seems fine. To me this metric should allow to track the  state machine for the lifecycle of an order.

Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:236]: @MartinquaXD mentioned that the current metric where `order_evets` table is used is going to be improved. But I don't have enough context in which way.
Currently, we have something like `ready->ready->ready->traded` where we can clearly see the time difference between the latest `ready` and `traded` events. With filtering `ready` events, the difference becomes bigger. For some of the subsequent labels, we can only update their timestamps. But for the `invalid` label, for example, we care about the initial timestamp rather than the latest one. Anyways, would need more context from Martin 🙂

Comment (fleupold) [reply crates/autopilot/src/solvable_orders.rs:236]: I'm pretty sure we always care about the initial timestamp (for ready -> ready -> traded, we should measure our performance from when your order first became ready).

Given Martin is off until the 8th feel free to merge either with or without the unification (but please follow up before closing the issue)

**Pull request merged**
